kernel/multicore_bringup/src/aarch64.rs

```diff
@@ -0,0 +1,214 @@
+use memory_aarch64::{read_mmu_config, asm_set_mmu_config_x2_x3};
+use kernel_config::memory::{PAGE_SIZE, KERNEL_STACK_SIZE_IN_PAGES};
+use psci::{cpu_on, error::Error::*};
+use zerocopy::FromBytes;
+use ap_start::kstart_ap;
+use volatile::Volatile;
+use core::arch::asm;
+use cpu::{CpuId, current_cpu, MpidrValue};
+use no_drop::NoDrop;
+
+/// The data items used when an AP core is booting up in ap_entry_point & ap_stage_two.
+#[cfg(target_arch = "aarch64")]
+#[derive(Debug, Default, FromBytes)]
+#[repr(C)]
+struct ApTrampolineData {
+    /// The ending virtual address (top) of the stack that was allocated for the new AP.
+    /// Will initially be zero before the BSP sets it to a real value
+    ap_stack_end: Volatile<VirtualAddress>,
+
+    /// The starting virtual address (bottom) of the stack that was allocated for the new AP.
+    ap_stack_start: Volatile<VirtualAddress>,
+
+    // Address to jump to after the MMU has been configured
+    ap_stage_two_virt_addr: Volatile<VirtualAddress>,
+
+    // Address used to access this structure once in ap_stage_two
+    ap_data_virt_addr: Volatile<VirtualAddress>,
+
+    // Boolean set to true by ap_stage_two when it's done using this structure
+    ap_ready: Volatile<u64>,
+
+    /// The physical address of the top-level P4 page table root.
+    ap_mmu_config: Volatile<PhysicalAddress>,
+}
+
+pub struct MulticoreBringupInfo;
+
+pub fn handle_ap_cores(
+    kernel_mmi_ref: &MmiRef,
+    _multicore_info: MulticoreBringupInfo,
+) -> Result<usize, &'static str> {
+    let mut online_cores = 0;
+
+    // This ApTrampolineData & MmuConfig will be read and written-to
+    // by all detected CPU cores, via both its physical
+    // and virtual addresses.
+    let mmu_config = read_mmu_config();
+    let mut ap_data: ApTrampolineData = Default::default();
+    ap_data.ap_data_virt_addr.write(VirtualAddress::new_canonical(&ap_data as *const _ as usize));
+    ap_data.ap_stage_two_virt_addr.write(VirtualAddress::new_canonical(ap_stage_two as usize));
+
+    let entry_point_phys_addr: PhysicalAddress;
+    let ap_data_phys_addr: PhysicalAddress;
+    let mut ap_startup_mapped_pages: MappedPages; // must be held throughout APs being booted up
+    {
+        let mut kernel_mmi = kernel_mmi_ref.lock();
+        let page_table = &mut kernel_mmi.page_table;
+
+        // get the physical address of ap_entry_point
+        let entry_point_virt_addr = VirtualAddress::new_canonical(ap_entry_point as usize);
+        entry_point_phys_addr = page_table.translate(entry_point_virt_addr).unwrap();
+
+        // get the physical address of the MmuConfig
+        let mmu_config_virt_addr = VirtualAddress::new_canonical(&mmu_config as *const _ as usize);
+        ap_data.ap_mmu_config.write(page_table.translate(mmu_config_virt_addr).unwrap());
+
+        // Note: this could be variable on aarch64
+        // we'd need a way to find any identity mappable page+frame range
+        const AP_STARTUP: usize = 0x10000;
```

Comment:
identity mapping shouldn't be too hard. What are the requirements; does it need to be at a certain address in low memory? For ex, on x86, it needs to be within the bounds of 16 bits, hence why this is so complicated.

Reply:
> identity mapping shouldn't be too hard. What are the requirements; does it need to be at a certain address in low memory? For ex, on x86, it needs to be within the bounds of 16 bits, hence why this is so complicated.

Nope; we only need an ID-mapping, whatever the address.